lib/core/include/irods/irods_logger.hpp

```diff
@@ -656,29 +663,34 @@ namespace irods::experimental::log
 
             [[nodiscard]] auto utc_timestamp() const -> std::string
             {
-                // clang-format off
-                using clock_type      = std::chrono::system_clock;
-                using time_point_type = std::chrono::time_point<clock_type>;
-                // clang-format on
-
-                timespec ts; // NOLINT(cppcoreguidelines-pro-type-member-init)
-
-                if (const auto ec = clock_gettime(CLOCK_REALTIME, &ts); ec != 0) {
-                    const auto now = clock_type::to_time_t(clock_type::now());
-                    std::stringstream utc_ss;
-                    utc_ss << std::put_time(std::gmtime(&now), "%FT%T.000Z"); // NOLINT(concurrency-mt-unsafe)
-                    return utc_ss.str();
+                namespace chrono = std::chrono;
+                using clock_type = chrono::system_clock;
+
+#if IRODS_CHRONO_FORMATTER_FMT
+                const auto tp = chrono::floor<chrono::milliseconds>(clock_type::now());
+                return fmt::format(FMT_COMPILE("{:%FT%H:%M:%S}Z"), tp);
+#else
+                char ts_buf[64]{};
+                std::timespec ts; // NOLINT(cppcoreguidelines-pro-type-member-init)
+                if (const int base = std::timespec_get(&ts, TIME_UTC); base != TIME_UTC) {
+                    if (const int ec = clock_gettime(CLOCK_REALTIME, &ts); ec != 0) {
+                        const auto tp = clock_type::now();
+                        const auto tp_s = chrono::floor<chrono::seconds>(tp);
+                        const auto tp_s_ms = chrono::floor<chrono::milliseconds>(tp_s);
+                        const auto tp_ms = chrono::floor<chrono::milliseconds>(tp) - tp_s_ms;
```

Comment:
How did we determine `floor` was needed?

Reply:
It was either that or `time_point_cast`; they both do effectively the same thing here. I did some timing tests and found that `floor` is marginally faster, at least on my machine.

Reply:
Cool cool.